File: hallucinate_app/python/hallucinate_app/graphrag.py
"""
GraphRAG Integration Layer

This module serves as an integration layer for the GraphRAG functionality
provided by the ipfs_datasets_py package. It does not implement any core 
functionality itself but provides standardized testing and access to the 
external module implementation.

The module's responsibility is to:
1. Import and provide access to GraphRAG from ipfs_datasets_py
2. Run comprehensive tests to ensure the GraphRAG functionality works
3. Integrate with the resource pool of the application
4. Provide a unified interface for other components to use GraphRAG
"""

import os
import json
import logging
import asyncio
import time
from pathlib import Path
from typing import Dict, List, Optional, Union, Any, Tuple

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger("graphrag_integration")

# Try to import IPFS Kit
try:
    from .ipfs_kit import IPFSKit, ipfs_kit
    has_ipfs_kit = True
except ImportError:
    logger.warning("Could not import IPFSKit, some functionality will be limited")
    has_ipfs_kit = False

# Try to import IPFS FAISS
try:
    from .ipfs_faiss import IPFSFaiss, ipfs_faiss
    has_ipfs_faiss = True
except ImportError:
    logger.warning("Could not import IPFSFaiss, vector search functionality will be limited")
    has_ipfs_faiss = False

# Try to import ipfs_datasets_py
try:
    import ipfs_datasets_py
    from ipfs_datasets_py import IPFSDatasetsPy
    has_ipfs_datasets_py = True
    
    # Check if GraphRAG is available in ipfs_datasets_py
    has_graphrag = hasattr(ipfs_datasets_py, 'GraphRAG')
    if not has_graphrag:
        logger.warning("GraphRAG not found in ipfs_datasets_py, functionality will be unavailable")
except ImportError:
    logger.warning("Could not import ipfs_datasets_py, GraphRAG functionality will be unavailable")
    has_ipfs_datasets_py = False
    has_graphrag = False

class GraphRAG:
    """
    Integration layer for GraphRAG functionality from ipfs_datasets_py
    """
    
    def __init__(self, resources=None, metadata=None):
        """
        Initialize GraphRAG integration layer
        
        Args:
            resources (dict): Resources required by GraphRAG
            metadata (dict): Metadata for operations
        """
        self.resources = resources or {}
        self.metadata = metadata or {}
        
        # Get storage directory from metadata or default
        self.storage_dir = self.metadata.get('storage_dir', os.path.join(
            os.path.expanduser('~'), '.cache', 'graphrag'
        ))
        os.makedirs(self.storage_dir, exist_ok=True)
        
        # Initialize modules
        if has_ipfs_kit:
            if 'ipfsKit' in self.resources:
                self.ipfs_kit = self.resources['ipfsKit']
            else:
                self.ipfs_kit = ipfs_kit
        else:
            self.ipfs_kit = None
        
        if has_ipfs_faiss:
            if 'ipfsFaiss' in self.resources:
                self.faiss = self.resources['ipfsFaiss']
            else:
                self.faiss = ipfs_faiss
        else:
            self.faiss = None
        
        # Initialize the GraphRAG implementation from ipfs_datasets_py
        self.graphrag_impl = None
        if has_graphrag:
            try:
                # Update resources for GraphRAG
                graphrag_resources = self.resources.copy()
                if self.ipfs_kit:
                    graphrag_resources['ipfsKit'] = self.ipfs_kit
                if self.faiss:
                    graphrag_resources['ipfsFaiss'] = self.faiss
                
                # Update metadata for GraphRAG
                graphrag_metadata = self.metadata.copy()
                graphrag_metadata['storage_dir'] = self.storage_dir
                
                # Create instance
                self.graphrag_impl = ipfs_datasets_py.GraphRAG(graphrag_resources, graphrag_metadata)
                logger.info("GraphRAG implementation initialized from ipfs_datasets_py")
            except Exception as e:
                logger.error(f"Failed to initialize GraphRAG from ipfs_datasets_py: {e}")
        else:
            logger.warning("GraphRAG implementation not available, using mock implementation")
        
        self.initialized = False
        self.version = "0.1"
        logger.info(f"GraphRAG integration layer initialized with storage={self.storage_dir}")
    
    async def init(self):
        """Initialize GraphRAG and its dependencies"""
        try:
            # Initialize modules
            if self.faiss:
                await self.faiss.init()
            
            # Initialize GraphRAG implementation
            if self.graphrag_impl:
                if hasattr(self.graphrag_impl, 'init'):
                    await self.graphrag_impl.init()
                    logger.info("GraphRAG implementation initialized")
                else:
                    logger.warning("GraphRAG implementation does not have init method")
            
            self.initialized = True
            return True
        except Exception as e:
            logger.error(f"GraphRAG initialization error: {e}")
            return False
    
    async def _forward_method(self, method_name, *args, **kwargs):
        """
        Forward method call to the GraphRAG implementation
        
        Args:
            method_name: Name of the method to call
            *args: Positional arguments to pass
            **kwargs: Keyword arguments to pass
            
        Returns:
            Result from the method or error dict
        """
        if not self.initialized:
            await self.init()
        
        try:
            if self.graphrag_impl:
                method = getattr(self.graphrag_impl, method_name, None)
                if method and callable(method):
                    return await method(*args, **kwargs)
                else:
                    logger.error(f"Method {method_name} not found in GraphRAG implementation")
                    return {"error": f"Method {method_name} not found in GraphRAG implementation"}
            else:
                logger.error("GraphRAG implementation not available")
                return {"error": "GraphRAG implementation not available"}
        except Exception as e:
            logger.error(f"Error calling {method_name}: {e}")
            return {"error": str(e)}
    
    # Forward standard operations to implementation
    
    async def add_document(self, document_id, text, metadata=None):
        """
        Add a document to the graph database
        
        Args:
            document_id: Unique ID for the document
            text: Document text content
            metadata: Optional metadata for the document
            
        Returns:
            Result from the add_document operation
        """
        return await self._forward_method("add_document", document_id, text, metadata)
    
    async def add_node(self, node_id, data, metadata=None, generate_embedding=True):
        """
        Add a node to the graph
        
        Args:
            node_id: Unique ID for the node
            data: Node data (text or other content)
            metadata: Optional metadata for the node
            generate_embedding: Whether to generate an embedding
            
        Returns:
            Result from the add_node operation
        """
        return await self._forward_method("add_node", node_id, data, metadata, generate_embedding)
    
    async def add_edge(self, source_id, target_id, weight=1.0, bidirectional=False):
        """
        Add an edge between nodes
        
        Args:
            source_id: Source node ID
            target_id: Target node ID
            weight: Edge weight
            bidirectional: Whether to add edges in both directions
            
        Returns:
            Result from the add_edge operation
        """
        return await self._forward_method("add_edge", source_id, target_id, weight, bidirectional)
    
    async def query(self, query_text, k=5, search_type="hybrid"):
        """
        Query the graph database
        
        Args:
            query_text: Query text
            k: Number of results to return
            search_type: Search type (vector, hybrid)
            
        Returns:
            Results from the query operation
        """
        return await self._forward_method("query", query_text, k, search_type)
    
    async def vector_search(self, query, k=5):
        """
        Perform vector similarity search
        
        Args:
            query: Query text or embedding
            k: Number of results to return
            
        Returns:
            Results from the vector search
        """
        return await self._forward_method("vector_search", query, k)
    
    async def hybrid_search(self, query, k=5, alpha=0.5):
        """
        Perform hybrid search combining vector similarity and graph traversal
        
        Args:
            query: Query text
            k: Number of results to return
            alpha: Balance between vector similarity and graph importance
            
        Returns:
            Results from the hybrid search
        """
        return await self._forward_method("hybrid_search", query, k, alpha)
    
    async def save_to_disk(self, path=None):
        """
        Save the graph database to disk
        
        Args:
            path: Path to save to or None for default
            
        Returns:
            Result from the save operation
        """
        return await self._forward_method("save_to_disk", path)
    
    async def load_from_disk(self, path):
        """
        Load the graph database from disk
        
        Args:
            path: Path to load from
            
        Returns:
            Result from the load operation
        """
        return await self._forward_method("load_from_disk", path)
    
    async def save_to_ipfs(self):
        """
        Save the graph database to IPFS
        
        Returns:
            Result from the IPFS save operation
        """
        return await self._forward_method("save_to_ipfs")
    
    async def load_from_ipfs(self, cid):
        """
        Load the graph database from IPFS
        
        Args:
            cid: IPFS CID for the database
            
        Returns:
            Result from the IPFS load operation
        """
        return await self._forward_method("load_from_ipfs", cid)
    
    async def get_stats(self):
        """
        Get statistics about the graph database
        
        Returns:
            Statistics dict
        """
        return await self._forward_method("get_stats")

    def test(self):
        """
        Run tests for the GraphRAG integration layer
        
        Returns:
            Dictionary with test results
        """
        try:
            # Run tests using asyncio
            import asyncio
            loop = asyncio.new_event_loop()
            
            results = {
                "success": False,
                "module": "graphrag_integration",
                "initialization": False,
                "implementation_found": self.graphrag_impl is not None,
                "document_operations": False,
                "node_operations": False,
                "vector_search": False,
                "hybrid_search": False,
                "persistence": False,
                "ipfs_integration": False,
                "capabilities": {
                    "graphrag_available": has_graphrag,
                    "ipfs_datasets_py_available": has_ipfs_datasets_py,
                    "faiss_available": self.faiss is not None,
                    "ipfs_available": self.ipfs_kit is not None
                }
            }
            
            # Test initialization
            init_result = loop.run_until_complete(self.init())
            results["initialization"] = init_result
            
            # If GraphRAG implementation is available, test its functionality
            if results["implementation_found"] and results["initialization"]:
                # Test GraphRAG's own test method if available
                if hasattr(self.graphrag_impl, 'test'):
                    impl_test_result = self.graphrag_impl.test()
                    results["implementation_test"] = impl_test_result
                    
                    # If implementation test succeeds, we can use those results
                    if impl_test_result.get("success", False):
                        for key in ["node_operations", "vector_search", "hybrid_search", "persistence", "ipfs_integration"]:
                            if key in impl_test_result:
                                results[key] = impl_test_result[key]
                        
                        results["success"] = True
                        return results
                
                # If no implementation test or it failed, run our own tests
                try:
                    # Test basic document operations
                    doc_id = f"test_doc_{int(time.time())}"
                    doc_result = loop.run_until_complete(
                        self.add_document(doc_id, "This is a test document for GraphRAG", {"test": True})
                    )
                    results["document_operations"] = doc_result is not None and not isinstance(doc_result, dict) and not "error" in str(doc_result)
                    
                    # Test node operations
                    if not results["document_operations"]:
                        # Try direct node operations
                        node_id = f"test_node_{int(time.time())}"
                        node_result = loop.run_until_complete(
                            self.add_node(node_id, "This is a test node", {"test": True})
                        )
                        results["node_operations"] = node_result is not None and not isinstance(node_result, dict) and not "error" in str(node_result)
                
                    # If we have nodes, test search
                    if results["node_operations"] or results["document_operations"]:
                        # Test vector search
                        vector_result = loop.run_until_complete(
                            self.vector_search("test document", 1)
                        )
                        results["vector_search"] = vector_result is not None and not isinstance(vector_result, dict) and not "error" in str(vector_result)
                        
                        # Test hybrid search
                        hybrid_result = loop.run_until_complete(
                            self.hybrid_search("test document", 1)
                        )
                        results["hybrid_search"] = hybrid_result is not None and not isinstance(hybrid_result, dict) and not "error" in str(hybrid_result)
                        
                        # Test persistence
                        save_result = loop.run_until_complete(
                            self.save_to_disk()
                        )
                        results["persistence"] = save_result is not None and not isinstance(save_result, dict) and not "error" in str(save_result)
                        
                        # Test IPFS if available
                        if self.ipfs_kit:
                            ipfs_result = loop.run_until_complete(
                                self.save_to_ipfs()
                            )
                            results["ipfs_integration"] = ipfs_result is not None and not isinstance(ipfs_result, dict) and not "error" in str(ipfs_result)
                except Exception as e:
                    logger.error(f"Error during GraphRAG testing: {e}")
            
            # Update overall success
            results["success"] = (
                results["initialization"] and 
                (results["node_operations"] or results["document_operations"]) and
                results["vector_search"]
            )
            
            return results
        except Exception as e:
            logger.error(f"GraphRAG test failed: {e}")
            return {
                "success": False,
                "module": "graphrag_integration",
                "error": str(e),
                "implementation_found": self.graphrag_impl is not None,
                "capabilities": {
                    "graphrag_available": has_graphrag,
                    "ipfs_datasets_py_available": has_ipfs_datasets_py,
                    "faiss_available": self.faiss is not None,
                    "ipfs_available": self.ipfs_kit is not None
                }
            }

# Create default instance
graphrag = GraphRAG()

File: hallucinate_app/node/graphrag.js
/**
 * GraphRAG Integration Layer
 *
 * This module serves as an integration layer for the GraphRAG functionality
 * provided by the ipfs_datasets_js package. It does not implement any core 
 * functionality itself but provides standardized testing and access to the 
 * external module implementation.
 *
 * The module's responsibility is to:
 * 1. Import and provide access to GraphRAG from ipfs_datasets_js
 * 2. Run comprehensive tests to ensure the GraphRAG functionality works
 * 3. Integrate with the resource pool of the application
 * 4. Provide a unified interface for other components to use GraphRAG
 */

// Import required modules
import fs from 'fs';
import path from 'path';
import os from 'os';
import { EventEmitter } from 'events';

// Configure logging
const logger = {
  info: (message) => console.log(`[INFO] [graphrag_integration] ${message}`),
  warning: (message) => console.warn(`[WARNING] [graphrag_integration] ${message}`),
  error: (message) => console.error(`[ERROR] [graphrag_integration] ${message}`)
};

// Try to import IPFS Kit
let ipfsKit;
let hasIpfsKit = false;
try {
  ipfsKit = (await import('./ipfs_kit.js')).default;
  hasIpfsKit = true;
} catch (error) {
  logger.warning(`Could not import IPFSKit, some functionality will be limited: ${error.message}`);
}

// Try to import IPFS FAISS
let ipfsFaiss;
let hasIpfsFaiss = false;
try {
  ipfsFaiss = (await import('./ipfs_faiss.js')).default;
  hasIpfsFaiss = true;
} catch (error) {
  logger.warning(`Could not import IPFSFaiss, vector search functionality will be limited: ${error.message}`);
}

// Try to import ipfs_datasets_js
let ipfsDatasetsJs;
let hasIpfsDatasetsJs = false;
let hasGraphrag = false;
try {
  ipfsDatasetsJs = await import('ipfs_datasets_js');
  hasIpfsDatasetsJs = true;
  
  // Check if GraphRAG is available in ipfs_datasets_js
  hasGraphrag = ipfsDatasetsJs.GraphRAG !== undefined;
  if (!hasGraphrag) {
    logger.warning('GraphRAG not found in ipfs_datasets_js, functionality will be unavailable');
  }
} catch (error) {
  logger.warning(`Could not import ipfs_datasets_js, GraphRAG functionality will be unavailable: ${error.message}`);
}

/**
 * Integration layer for GraphRAG functionality from ipfs_datasets_js
 */
export class GraphRAG extends EventEmitter {
  /**
   * Initialize GraphRAG integration layer
   * 
   * @param {Object} resources - Resources required by GraphRAG
   * @param {Object} metadata - Metadata for operations
   */
  constructor(resources = null, metadata = null) {
    super();
    this.resources = resources || {};
    this.metadata = metadata || {};
    
    // Get storage directory from metadata or default
    this.storageDir = this.metadata.storageDir || path.join(
      os.homedir(), '.cache', 'graphrag'
    );
    fs.mkdirSync(this.storageDir, { recursive: true });
    
    // Initialize modules
    if (hasIpfsKit) {
      if (this.resources.ipfsKit) {
        this.ipfsKit = this.resources.ipfsKit;
      } else {
        this.ipfsKit = ipfsKit;
      }
    } else {
      this.ipfsKit = null;
    }
    
    if (hasIpfsFaiss) {
      if (this.resources.ipfsFaiss) {
        this.faiss = this.resources.ipfsFaiss;
      } else {
        this.faiss = ipfsFaiss;
      }
    } else {
      this.faiss = null;
    }
    
    // Initialize the GraphRAG implementation from ipfs_datasets_js
    this.graphragImpl = null;
    if (hasGraphrag) {
      try {
        // Update resources for GraphRAG
        const graphragResources = { ...this.resources };
        if (this.ipfsKit) {
          graphragResources.ipfsKit = this.ipfsKit;
        }
        if (this.faiss) {
          graphragResources.ipfsFaiss = this.faiss;
        }
        
        // Update metadata for GraphRAG
        const graphragMetadata = { ...this.metadata };
        graphragMetadata.storageDir = this.storageDir;
        
        // Create instance
        this.graphragImpl = new ipfsDatasetsJs.GraphRAG(graphragResources, graphragMetadata);
        logger.info('GraphRAG implementation initialized from ipfs_datasets_js');
      } catch (error) {
        logger.error(`Failed to initialize GraphRAG from ipfs_datasets_js: ${error}`);
      }
    } else {
      logger.warning('GraphRAG implementation not available, using mock implementation');
    }
    
    this.initialized = false;
    this.version = '0.1';
    logger.info(`GraphRAG integration layer initialized with storage=${this.storageDir}`);
  }
  
  /**
   * Initialize GraphRAG and its dependencies
   * @returns {Promise<boolean>} True if initialization successful
   */
  async init() {
    try {
      // Initialize modules
      if (this.faiss) {
        await this.faiss.init();
      }
      
      // Initialize GraphRAG implementation
      if (this.graphragImpl) {
        if (typeof this.graphragImpl.init === 'function') {
          await this.graphragImpl.init();
          logger.info('GraphRAG implementation initialized');
        } else {
          logger.warning('GraphRAG implementation does not have init method');
        }
      }
      
      this.initialized = true;
      return true;
    } catch (error) {
      logger.error(`GraphRAG initialization error: ${error}`);
      return false;
    }
  }
  
  /**
   * Forward method call to the GraphRAG implementation
   * 
   * @param {string} methodName - Name of the method to call
   * @param  {...any} args - Arguments to pass to the method
   * @returns {Promise<any>} Result from the method or error object
   * @private
   */
  async _forwardMethod(methodName, ...args) {
    if (!this.initialized) {
      await this.init();
    }
    
    try {
      if (this.graphragImpl) {
        const method = this.graphragImpl[methodName];
        if (method && typeof method === 'function') {
          return await method.apply(this.graphragImpl, args);
        } else {
          logger.error(`Method ${methodName} not found in GraphRAG implementation`);
          return { error: `Method ${methodName} not found in GraphRAG implementation` };
        }
      } else {
        logger.error('GraphRAG implementation not available');
        return { error: 'GraphRAG implementation not available' };
      }
    } catch (error) {
      logger.error(`Error calling ${methodName}: ${error}`);
      return { error: error.toString() };
    }
  }
  
  // Forward standard operations to implementation
  
  /**
   * Add a document to the graph database
   * 
   * @param {string} documentId - Unique ID for the document
   * @param {string} text - Document text content
   * @param {Object} metadata - Optional metadata for the document
   * @returns {Promise<any>} Result from the add_document operation
   */
  async addDocument(documentId, text, metadata = null) {
    return await this._forwardMethod('addDocument', documentId, text, metadata);
  }
  
  /**
   * Add a node to the graph
   * 
   * @param {string} nodeId - Unique ID for the node
   * @param {string|Object} data - Node data (text or other content)
   * @param {Object} metadata - Optional metadata for the node
   * @param {boolean} generateEmbedding - Whether to generate an embedding
   * @returns {Promise<any>} Result from the add_node operation
   */
  async addNode(nodeId, data, metadata = null, generateEmbedding = true) {
    return await this._forwardMethod('addNode', nodeId, data, metadata, generateEmbedding);
  }
  
  /**
   * Add an edge between nodes
   * 
   * @param {string} sourceId - Source node ID
   * @param {string} targetId - Target node ID
   * @param {number} weight - Edge weight
   * @param {boolean} bidirectional - Whether to add edges in both directions
   * @returns {Promise<any>} Result from the add_edge operation
   */
  async addEdge(sourceId, targetId, weight = 1.0, bidirectional = false) {
    return await this._forwardMethod('addEdge', sourceId, targetId, weight, bidirectional);
  }
  
  /**
   * Query the graph database
   * 
   * @param {string} queryText - Query text
   * @param {number} k - Number of results to return
   * @param {string} searchType - Search type (vector, hybrid)
   * @returns {Promise<any>} Results from the query operation
   */
  async query(queryText, k = 5, searchType = 'hybrid') {
    return await this._forwardMethod('query', queryText, k, searchType);
  }
  
  /**
   * Perform vector similarity search
   * 
   * @param {string|Array} query - Query text or embedding
   * @param {number} k - Number of results to return
   * @returns {Promise<any>} Results from the vector search
   */
  async vectorSearch(query, k = 5) {
    return await this._forwardMethod('vectorSearch', query, k);
  }
  
  /**
   * Perform hybrid search combining vector similarity and graph traversal
   * 
   * @param {string} query - Query text
   * @param {number} k - Number of results to return
   * @param {number} alpha - Balance between vector similarity and graph importance
   * @returns {Promise<any>} Results from the hybrid search
   */
  async hybridSearch(query, k = 5, alpha = 0.5) {
    return await this._forwardMethod('hybridSearch', query, k, alpha);
  }
  
  /**
   * Save the graph database to disk
   * 
   * @param {string} path - Path to save to or null for default
   * @returns {Promise<any>} Result from the save operation
   */
  async saveToDisk(path = null) {
    return await this._forwardMethod('saveToDisk', path);
  }
  
  /**
   * Load the graph database from disk
   * 
   * @param {string} path - Path to load from
   * @returns {Promise<any>} Result from the load operation
   */
  async loadFromDisk(path) {
    return await this._forwardMethod('loadFromDisk', path);
  }
  
  /**
   * Save the graph database to IPFS
   * 
   * @returns {Promise<any>} Result from the IPFS save operation
   */
  async saveToIpfs() {
    return await this._forwardMethod('saveToIpfs');
  }
  
  /**
   * Load the graph database from IPFS
   * 
   * @param {string} cid - IPFS CID for the database
   * @returns {Promise<any>} Result from the IPFS load operation
   */
  async loadFromIpfs(cid) {
    return await this._forwardMethod('loadFromIpfs', cid);
  }
  
  /**
   * Get statistics about the graph database
   * 
   * @returns {Promise<any>} Statistics object
   */
  async getStats() {
    return await this._forwardMethod('getStats');
  }

  /**
   * Run tests for the GraphRAG integration layer
   * 
   * @returns {Object} Object with test results
   */
  test() {
    const runAsync = async () => {
      try {
        const results = {
          success: false,
          module: 'graphrag_integration',
          initialization: false,
          implementationFound: this.graphragImpl !== null,
          documentOperations: false,
          nodeOperations: false,
          vectorSearch: false,
          hybridSearch: false,
          persistence: false,
          ipfsIntegration: false,
          capabilities: {
            graphragAvailable: hasGraphrag,
            ipfsDatasetsJsAvailable: hasIpfsDatasetsJs,
            faissAvailable: this.faiss !== null,
            ipfsAvailable: this.ipfsKit !== null
          }
        };
        
        // Test initialization
        const initResult = await this.init();
        results.initialization = initResult;
        
        // If GraphRAG implementation is available, test its functionality
        if (results.implementationFound && results.initialization) {
          // Test GraphRAG's own test method if available
          if (typeof this.graphragImpl.test === 'function') {
            const implTestResult = this.graphragImpl.test();
            results.implementationTest = implTestResult;
            
            // If implementation test succeeds, we can use those results
            if (implTestResult && implTestResult.success) {
              for (const key of ['nodeOperations', 'vectorSearch', 'hybridSearch', 'persistence', 'ipfsIntegration']) {
                if (key in implTestResult) {
                  results[key] = implTestResult[key];
                }
              }
              
              results.success = true;
              return results;
            }
          }
          
          // If no implementation test or it failed, run our own tests
          try {
            // Test basic document operations
            const docId = `test_doc_${Date.now()}`;
            const docResult = await this.addDocument(docId, 'This is a test document for GraphRAG', { test: true });
            results.documentOperations = docResult !== null && 
              typeof docResult !== 'object' && 
              !(docResult && docResult.error);
            
            // Test node operations if document operations failed
            if (!results.documentOperations) {
              // Try direct node operations
              const nodeId = `test_node_${Date.now()}`;
              const nodeResult = await this.addNode(nodeId, 'This is a test node', { test: true });
              results.nodeOperations = nodeResult !== null && 
                typeof nodeResult !== 'object' && 
                !(nodeResult && nodeResult.error);
            }
          
            // If we have nodes, test search
            if (results.nodeOperations || results.documentOperations) {
              // Test vector search
              const vectorResult = await this.vectorSearch('test document', 1);
              results.vectorSearch = vectorResult !== null && 
                typeof vectorResult !== 'object' && 
                !(vectorResult && vectorResult.error);
              
              // Test hybrid search
              const hybridResult = await this.hybridSearch('test document', 1);
              results.hybridSearch = hybridResult !== null && 
                typeof hybridResult !== 'object' && 
                !(hybridResult && hybridResult.error);
              
              // Test persistence
              const saveResult = await this.saveToDisk();
              results.persistence = saveResult !== null && 
                typeof saveResult !== 'object' && 
                !(saveResult && saveResult.error);
              
              // Test IPFS if available
              if (this.ipfsKit) {
                const ipfsResult = await this.saveToIpfs();
                results.ipfsIntegration = ipfsResult !== null && 
                  typeof ipfsResult !== 'object' && 
                  !(ipfsResult && ipfsResult.error);
              }
            }
          } catch (error) {
            logger.error(`Error during GraphRAG testing: ${error}`);
          }
        }
        
        // Update overall success
        results.success = (
          results.initialization && 
          (results.nodeOperations || results.documentOperations) &&
          results.vectorSearch
        );
        
        return results;
      } catch (error) {
        logger.error(`GraphRAG test failed: ${error}`);
        return {
          success: false,
          module: 'graphrag_integration',
          error: error.toString(),
          implementationFound: this.graphragImpl !== null,
          capabilities: {
            graphragAvailable: hasGraphrag,
            ipfsDatasetsJsAvailable: hasIpfsDatasetsJs,
            faissAvailable: this.faiss !== null,
            ipfsAvailable: this.ipfsKit !== null
          }
        };
      }
    };

    // Run async tests and return results synchronously
    let testResults;
    runAsync().then(results => {
      testResults = results;
    }).catch(error => {
      testResults = {
        success: false,
        module: 'graphrag_integration',
        error: error.toString()
      };
    });

    // Return dummy results immediately (actual results will be emitted)
    return {
      success: false,
      module: 'graphrag_integration',
      message: 'Tests are running asynchronously, results will be emitted as events',
      async: true
    };
  }
}

// Create default instance
const graphrag = new GraphRAG();

export default graphrag;